---
title: "Comparative heatmaps of samples (master files)"
author: "Ellen Dimmen Chapple"
date: "2023-08-25"
output: 
  html_document:
    code_folding: hide
    toc: true
    number_sections: true
    toc_float: true
    theme: paper
editor_options:
  chunk_output_type: console
---

# Libraries and lists

```{r Packages, message=FALSE, warning=FALSE}

library(tidyverse)
library(ggplot2)
library(ComplexHeatmap)
library(tibble)
library(cowplot)
library(gdata)
library(circlize)
library(ggtree)
library(gridExtra)
library(paletteer)
library(forcats)
library(ggpubr)
library(ggridges)

# Run version of ComPlEx
VER <- "V4" # first run: "" second run: "-V2"

```



```{r name list}

species_list <- c("Lodge", "Asp", "Nor","Scots","Birch","Cher")
#"Asp", "Nor", ,  "Cher"
combo <- data.frame(t(combn(species_list, 2)))

```



```{r file names}

# List of file names.

file_list_1 <- c()
for (i in 1:nrow(combo)){
  
  s1 <- combo[i, "X1"]
  s2 <- combo[i, "X2"]
  
  file_name <- paste0("Data/DATA/comparisonFiles/comparison-",s1,"_",s2,"-pearsonMR0.03no-table-",VER,".RData")
  file_list_1 <- append(file_list_1, file_name, after = length(file_list_1))
    
}

file_list_2 <- c()
for (i in species_list){
  
  expr_name <- paste0("Data/DATA/transcriptomicsData/",i,"Wood_transcriptomics.txt.gz")
  file_list_2 <- append(file_list_2, expr_name, after = length(file_list_2))
    
}


```

# Heatmaps

## Correlation matrices

```{r corr matrix}

pval <- 0.1

for (x in file_list_1){
  if (file.exists(x)){ 

    # x <- file_list_1[6]
    load(x) 

    
    co_expressologs <- comparison_table%>%
    select(Species1, Species2,OrthoGroup, Max.p.val) %>%
    mutate_at("Max.p.val", as.numeric) %>%
    filter(Max.p.val < pval) # Only co-expressologs


    co_expressologs <- co_expressologs %>% 
    ungroup() %>%
    arrange(Max.p.val) %>% 
    distinct(Species1,  .keep_all = T) %>%
    distinct( Species2, .keep_all = T) # Each gene can only appear in one co-expressolog

    # SPECIES 1
    key_word_s1 <- sapply(strsplit(x, "_"), "[",1) 
    key_word_s1 <- sapply(strsplit(key_word_s1, "-"), "[",2) 


    for (i in file_list_2){ 
      if (grepl(key_word_s1, i)){ # If the string above contains key_word_1 ...
      
        # i <- file_list_2[2]
        expr_data <- read.delim(i) # ... read expression file.
      
    
      s1_expression <- co_expressologs %>% 
      left_join(expr_data, join_by(Species1 == Genes)) %>%
      select(-c("Species2", "Max.p.val", "OrthoGroup")) %>%
      as_tibble() %>%
      column_to_rownames(var = "Species1") %>%
      select(contains("1."))
      
      s1_expression[is.na(s1_expression)] <- 0
      
      }}
    
    
    
    # SPECIES 2
    key_word_s2 <- sapply(strsplit(x, "_"), "[",2)
    key_word_s2 <- sapply(strsplit(key_word_s2, "-"), "[",1) 

    for (i in file_list_2){ 
      if (grepl(key_word_s2, i)){
        
        # i <- file_list_2[3]
      expr_data <- read.delim(i) 
      
      s2_expression <- co_expressologs %>% 
      left_join(expr_data, join_by(Species2 == Genes)) %>%
      select(-c("Species1", "Max.p.val", "OrthoGroup")) %>%
      as_tibble() %>%
      column_to_rownames(var = "Species2") %>%
      select(contains("1."))
      
      s2_expression[is.na(s2_expression)] <- 0
      
  }}
    
    # Correlation matrix
    corr_matrix <- round(cor(s1_expression,s2_expression, method = "pearson"), 4) 

    new_name <- as.character(paste0("corr_",key_word_s1, "-",key_word_s2))
    mv("corr_matrix", new_name) 
    print(new_name)

  }
  
  }

# Transposing matrices

to_transpose <- c("corr_Asp-Scots", "corr_Asp-Nor", "corr_Asp-Cher", "corr_Asp-Birch", "corr_Birch-Cher", "corr_Nor-Scots")

for(x in to_transpose){

  # x <- to_transpose[1]
    key_word_s1 <- sapply(strsplit(x, "_"), "[",2)
    key_word_s1 <- sapply(strsplit(key_word_s1, "-"), "[",1)
    key_word_s2 <- sapply(strsplit(x, "-"), "[",2) 

    corr_t <- t(get(x))
    new_name <- as.character(paste0("corr_",key_word_s2, "-",key_word_s1))
    mv("corr_t", new_name) 
    rm(list = x)
    
   print(x) 

}

# Self-correlation matrices

load("Data/DATA/all_expressed_genes.RData")
co_expressologs_all <- expr_genes %>% 
  filter(Max.p.Val < 0.1)

for(i in file_list_2){

  # i <- file_list_2[2]
  
  species <- sapply(strsplit(i, "transcriptomicsData/"), "[",2)
  species <- sapply(strsplit(species, "Wood"), "[",1)
  print(species)

  
  co_expressologs_one_species <- co_expressologs_all %>% 
    filter(species %in% Species1 | species %in% Species2)
  
  col_1 <-co_expressologs_one_species %>% 
    select(Species1, GeneSpecies1) %>% 
    rename(Species = "Species1", Genes = "GeneSpecies1") %>% 
    filter(Species == species)
  
  col_2 <-co_expressologs_one_species %>% 
    select(Species2, GeneSpecies2) %>% 
    rename(Species = "Species2", Genes = "GeneSpecies2")%>% 
    filter(Species == species)
  
  if(nrow(col_1) == 0) {
    combined_df <- col_2
  }
  
  
  if(nrow(col_2) == 0) {
    combined_df <- col_1
  }

  if(nrow(col_2) > 0 & nrow(col_1) > 0){
    combined_df <- rbind(col_1, col_2)
  }
  
  combined_df <- combined_df %>% 
    distinct(Genes, .keep_all = T)
    
  
  expression <- read.delim(i) %>%
    filter(Genes %in% combined_df$Genes) %>% 
    select(contains("1."))

  
  
  

  corr_matrix <- round(cor(expression[,-1], method = "pearson"), 4)
  new_name <- as.character(paste0("corr_",species, "-",species))
  mv("corr_matrix", new_name) # rename the heatmaps for plotting in grid.

}


# ---------------------------------  Individual heatmaps--------------------------------- 


corr_list <- c("corr_Cher-Asp", "corr_Birch-Asp", "corr_Lodge-Asp", "corr_Nor-Asp", "corr_Scots-Asp", "corr_Cher-Birch", "corr_Nor-Birch", "corr_Nor-Cher", "corr_Scots-Birch", "corr_Scots-Cher", "corr_Scots-Nor", "corr_Lodge-Birch", "corr_Lodge-Cher", "corr_Lodge-Nor", "corr_Lodge-Scots", "corr_Scots-Scots", "corr_Nor-Nor", "corr_Lodge-Lodge", "corr_Asp-Asp", "corr_Birch-Birch", "corr_Cher-Cher"
)

# All on same sep legend

for(matrix in corr_list){


    key_word_s1 <- sapply(strsplit(matrix, "_"), "[",2)
    key_word_s1 <- sapply(strsplit(key_word_s1, "-"), "[",1)
    key_word_s2 <- sapply(strsplit(matrix, "-"), "[",2)

    lowest <- min(get(matrix))
    highest <- max(get(matrix))
    between <- median(get(matrix))
    
    col_function_1 <- colorRamp2(c(lowest,  between, highest), c("#800026", "#fd8d3c", "#ffffcc"))
    # col_function_1 <- colorRamp2(c(lowest,  between, highest), c("#0c2c84", "#41b6c4", "#ffffd9"))
    # col_function_1 <- colorRamp2(c(lowest,  between, highest), c("#005a32", "#78c679", "#ffffe9"))
    # col_function_1 <- colorRamp2(c(lowest,  between, highest), c("#440154", "#316880","#dcebf2")) #FFB000
    # col_function_1 <- colorRamp2(c(lowest,  between, highest), c( "black", "grey66","white"))
    # col_function_1 <- colorRamp2(c(lowest,  between, highest), c("#440154", "#3B2F5EFF","#38AAACFF"))
    # col_function_1 <- colorRamp2(c(0,  0.5, 0.9), c("#440154", "#316880","#fde725"))
    print(matrix)
  

   hm <-  Heatmap(get(matrix),show_column_names = FALSE,
                  show_row_names = FALSE,
    width = ncol(get(matrix))*unit(5, "mm"),
    height = nrow(get(matrix))*unit(5, "mm"),
    row_order = 1:nrow(get(matrix)),
    column_order = 1:ncol(get(matrix)), col = col_function_1, show_heatmap_legend = F
               #         ,heatmap_width = unit(3, "cm"),
               # heatmap_height = unit(3, "cm")
    )




    new_name <- as.character(paste0("hm-",key_word_s1, "-",key_word_s2))
    mv("hm", new_name) # rename the heatmaps for plotting in grid.
    
}

g1 <- grid.grabExpr(print(`hm-Birch-Asp`))
plot_grid(g1)

g2 <- grid.grabExpr(print(`hm-Cher-Asp`)) #
g3 <- grid.grabExpr(print(`hm-Cher-Birch`)) #
g4 <- grid.grabExpr(print(`hm-Nor-Asp`)) #
g5 <- grid.grabExpr(print(`hm-Nor-Birch`)) 
g6 <- grid.grabExpr(print(`hm-Nor-Cher`))
g7 <- grid.grabExpr(print(`hm-Scots-Asp`)) #
g8 <- grid.grabExpr(print(`hm-Scots-Birch`))
g9 <- grid.grabExpr(print(`hm-Scots-Cher`))
g10 <- grid.grabExpr(print(`hm-Scots-Nor`))  
g11 <- grid.grabExpr(print(`hm-Lodge-Asp`))
g12 <- grid.grabExpr(print(`hm-Lodge-Birch`))
g13 <- grid.grabExpr(print(`hm-Lodge-Cher`))
g14 <- grid.grabExpr(print(`hm-Lodge-Nor`))
g15 <- grid.grabExpr(print(`hm-Lodge-Scots`))
# Diag
g16 <- grid.grabExpr(print(`hm-Asp-Asp`))
g17 <- grid.grabExpr(print(`hm-Birch-Birch`))
g18 <- grid.grabExpr(print(`hm-Cher-Cher`))
g19 <- grid.grabExpr(print(`hm-Nor-Nor`))
g20 <- grid.grabExpr(print(`hm-Scots-Scots`))
g21 <- grid.grabExpr(print(`hm-Lodge-Lodge`))

plot_grid(g2)

 hm <-plot_grid(g16,NULL, NULL, NULL, NULL,NULL,
          g1, g17, NULL, NULL,NULL, NULL,
          g2, g3, g18, NULL, NULL, NULL,
          g4, g5, g6, g19, NULL, NULL,
          g7, g8, g9, g10, g20, NULL,
          g11, g12, g13, g14, g15, g21, nrow = 6, ncol = 6)

 hm
  
#  hm <-plot_grid(g1,  NULL, NULL,NULL, NULL,
#           g2, g3,  NULL, NULL, NULL,
#           g4, g5, g6,  NULL, NULL,
#           g7, g8, g9, g10,  NULL,
#           g11, g12, g13, g14, g15,  nrow = 5, ncol = 5)
# hm

 
```


## Density plots

```{r }

df_for_density <- c()
# palette_1 <- c("#440154", "#316880","#fde725")

for(matrix in corr_list){
  
  # matrix <- corr_list[2]
  
  pair <- sapply(strsplit(matrix, "_"), "[",2)
  correlation_vect <- c(get(matrix))
  
  df <- data.frame(
    pair = rep(pair, length(correlation_vect)),
    correlation_values = correlation_vect
  )
  
  df_for_density <- rbind(df_for_density, df)
  
  # print(paste0(pair, ": ", length(correlation_vect)))
  
}


angio_pairs <- data.frame(pair = c("Cher-Asp" ,  "Birch-Asp", "Cher-Birch" ))
gymno_pairs <- data.frame(pair = c(  "Lodge-Nor" , "Lodge-Scots", "Scots-Nor"))
self <- data.frame(pair = c(   "Scots-Scots" ,"Nor-Nor"   ,  "Lodge-Lodge", "Asp-Asp" ,    "Birch-Birch", "Cher-Cher"  ))


# palette_1 <- c("#440154","#edc0cd","#994c9c")
# palette_1 <-  c("#06282c", "#448c8a","#eeffee")

all_non_self <- df_for_density %>%
  filter( !(pair %in% self$pair))

all_non_self[all_non_self == "Lodge-Scots"] <- "P.contorta - P.sylvestris"
all_non_self[all_non_self == "Lodge-Nor"] <- "P.contorta - P.abies"
all_non_self[all_non_self == "Scots-Nor"] <- "P.sylvestris - P.abies"
all_non_self[all_non_self == "Lodge-Cher"] <- "P.contorta - P.avium"
all_non_self[all_non_self == "Lodge-Birch"] <- "P.contorta - B.pendula"
all_non_self[all_non_self == "Scots-Cher"] <- "P.sylvestris - P.avium"
all_non_self[all_non_self == "Scots-Birch"] <- "P.sylvestris - B.pendula"
all_non_self[all_non_self == "Nor-Cher"] <- "P.abies - P.avium"
all_non_self[all_non_self == "Nor-Birch"] <- "P.abies - B.pendula"
all_non_self[all_non_self == "Scots-Asp"] <- "P.sylvestris - P.tremula"
all_non_self[all_non_self == "Nor-Asp"] <- "P.abies - P.tremula"
all_non_self[all_non_self == "Lodge-Asp"] <- "P.contorta - P.tremula"
all_non_self[all_non_self == "Cher-Birch"] <- "P.avium - B.pendula"
all_non_self[all_non_self == "Birch-Asp"] <- "B.pendula - P.tremula"
all_non_self[all_non_self == "Cher-Asp"] <- "P.avium - P.tremula"

speciesPairs <- tibble(SpeciesPair = all_non_self %>% distinct(pair) %>% pull(pair),
                       SpeciesPairClade = c("Angiosperm", "Angiosperm", "Crossed", "Crossed", "Crossed", "Angiosperm", "Crossed", "Crossed",
                       "Crossed", "Crossed", "Gymnosperm", "Crossed", "Crossed", "Gymnosperm", "Gymnosperm"))


all_non_self <-  left_join(all_non_self , speciesPairs, by = join_by("pair" == "SpeciesPair" ))

all_non_self <- all_non_self%>%
  mutate(pair = fct_relevel(pair,  "P.avium - P.tremula", "B.pendula - P.tremula", "P.avium - B.pendula",  "P.contorta - P.tremula" ,  "P.abies - P.tremula",     "P.sylvestris - P.tremula",   "P.abies - B.pendula"  ,   "P.abies - P.avium","P.sylvestris - B.pendula", "P.sylvestris - P.avium"  , "P.contorta - B.pendula", "P.contorta - P.avium",  "P.sylvestris - P.abies"  ,"P.contorta - P.abies", "P.contorta - P.sylvestris"))


angio_density <- all_non_self %>% 
  filter(SpeciesPairClade == "Angiosperm") %>% 
  group_by(pair) 

gymno_density <- all_non_self %>% 
  filter(SpeciesPairClade == "Gymnosperm") %>% 
  group_by(pair)

crossed_density <- all_non_self %>% 
  filter(SpeciesPairClade == "Crossed") %>% 
  group_by(pair)

# Line plot - solid colour
# 
# p1 <-  ggplot(angio_density, aes(x = correlation_values, groups = pair)) + geom_density(linewidth = 2, colour = "#ffc000") + labs(title = "Angiosperm pairs",x = "", y = "Density")  + theme_classic()+  ylim(0,4)+ xlim(0,1)+
#   theme(plot.title = element_text(hjust = 1, size = 20, face = "bold"), axis.title = element_text(size = 20), axis.text =  element_text(size = 15), axis.title.x = element_text(hjust = 0.5, vjust = -1)) 
# 
# 
# p2 <-ggplot(gymno_density, aes(x = correlation_values, groups = pair)) + geom_density(linewidth = 2, colour =   "#cf4447") + labs(title = "Gymnosperm pairs",x = "", y = "Density") +  ylim(0,4)+ xlim(0,1) + theme_classic()+  
#   theme(plot.title = element_text(hjust = 1, size = 20, face = "bold"), axis.title = element_text(size = 20), axis.text =  element_text(size = 15), axis.title.x = element_text(hjust = 0.5, vjust = -1)) 
# 
# 
# p3 <-ggplot(crossed_density, aes(x = correlation_values, groups = pair)) + geom_density(linewidth = 2, colour = "#0c3537") + labs( title = "Crossed pairs",x = "Correlation", y = "Density") +  ylim(0,4)+ xlim(0,1)+ theme_classic()+  
#   theme(plot.title = element_text(hjust = 1, size = 20, face = "bold"), axis.title = element_text(size = 20), axis.text =  element_text(size = 15), axis.title.x = element_text(hjust = 0.5, vjust = -2)) 
# 
# p1
# p2
# p3

# Line plot - lines
title_size <- 25
axis_title_size <- 20
axis_text_size <- 20

p1 <-  ggplot(angio_density, aes(x = correlation_values, groups = pair)) + geom_density(linewidth = 2, linetype = "dotted") + labs(title = "Angiosperm pairs",x = "", y = "Density") +  ylim(0,4)+ xlim(0,1) +  theme_classic()+ 
  theme(plot.title = element_text(hjust = 1, size = title_size, colour = "firebrick"), axis.title = element_text(size = axis_title_size, vjust = 2.5), axis.text =  element_text(size = axis_title_size))
p1

p2 <-ggplot(gymno_density, aes(x = correlation_values, groups = pair)) + geom_density(linewidth = 2) + labs(title = "Gymnosperm pairs",x = "", y = "Density") +  ylim(0,4)+ xlim(0,1) + 
  theme_classic()+ 
  theme(plot.title = element_text(hjust = 1, size = title_size, colour = "firebrick"), axis.title = element_text(size = axis_title_size, vjust = 2.5), axis.text =  element_text(size = axis_title_size))

p2

p3 <-ggplot(crossed_density, aes(x = correlation_values, groups = pair)) + geom_density(linewidth = 1) + labs( title = "Crossed pairs",x = "Correlation", y = "Density") +  ylim(0,4)+ xlim(0,1) + theme_classic()+
  theme(plot.title = element_text(hjust = 1, size = title_size, colour = "firebrick"), axis.title = element_text(size = axis_title_size, vjust = 2.5), axis.text =  element_text(size = axis_title_size))


p3


```


```{r}

# Plotting comparison heatmaps with phylogenetic trees
library(patchwork)
library(magick)


tree_full_name <- read.tree(text ="( ((Lodgepole pine, Scots pine), (Norway spruce)), ((Cherry,  Birch),(Aspen))  );")
tree_full_name_mirror <- read.tree(text ="( ((Aspen),(Birch, Cherry)), ((Norway spruce),(Scots pine, Lodgepole pine)));")

tree_full_name_lat <- read.tree(text ="( ((P.contorta, P.sylvestris), (P.abies)), ((P.avium,  B.pendula),(P.tremula))  );")
tree_full_name_mirror_lat <- read.tree(text ="( ((P.tremula),(B.pendula, P.avium)), ((P.abies),(P.sylvestris , P.contorta)));")

# tree_short_name <- read.tree(text ="( ((LP, SP), (NS)), ((C, B),(A))  );")
# tree_short_name_mirror <- read.tree(text ="( ((A),(B, C)), ((NS),(SP, LP)));")

tree_grouped <- groupClade(tree_full_name_mirror_lat, .node = c(8))


# ---- Trees with coloured clades -----

# tree_1 <-ggtree(tree_grouped, aes(color = group),ladderize = F, size = 1) + 
#   scale_color_manual(values=c("#cf4447", "#ffc000", "grey")) + geom_tiplab(color = "grey9",size = 5, fontface = "italic",vjust = -2, hjust = 0.5) + xlim(-10, 10) + theme(legend.position = "none") 
# tree_1
# 
# tree_horo <- tree_1 + rotate() 
# tree_grouped_vert <- groupClade(tree_full_name_lat, .node = c(8))
# tree_vert <- ggtree(tree_grouped_vert, aes(color = group), ladderize = F, size = 1) + geom_tiplab(color = "grey9",size = 5, angle = 0, fontface = "italic", hjust = -0.1)  + xlim(-10, 10) +scale_color_manual(values=c("#ffc000","#cf4447",  "grey")) + theme(legend.position = "none")
# 
# tree_vert

# ----- Trees with pattern/no pattern clades ------

lines = c(rep("dotted", 3), rep("solid", 3), rep("dotted", 4),  rep("solid", 3))

tree_1 <-ggtree(tree_grouped, linetype = lines ,ladderize = F, size = 1) + geom_tiplab(color = "grey9",size = 5, fontface = "italic",vjust = -2, hjust = 0.5) + xlim(-10, 10) + theme(legend.position = "none") 
tree_1

tree_horo <- tree_1 + rotate() 


tree_grouped_vert <- groupClade(tree_full_name_lat, .node = c(8))
tree_vert <- ggtree(tree_grouped_vert, linetype = rev(lines) , ladderize = F, size = 1) + geom_tiplab(color = "grey9",size = 5, angle = 0, fontface = "italic", hjust = -0.1)  + xlim(-10, 10)  + theme(legend.position = "none")

tree_vert


# Just for saving

# tree_grouped <- groupClade(tree_full_name_lat, .node = c(8))
# tree_upside_down <- ggtree(tree_grouped, linetype = rev(lines) ,ladderize = F, size = 1) + geom_tiplab(color = "grey9",size = 5, fontface = "italic",vjust = 2, hjust = 0.5, angle = 180) + xlim(-10, 10) + theme(legend.position = "none") + rotate()

```




```{r}
# Colour bar reflecting low and high correlation
col_fun <- colorRamp2(c(0, 0.5, 1), c("#800026", "#fd8d3c", "#ffffcc"))
# col_fun <- colorRamp2(c(0, 0.5, 1), c("#0c2c84", "#41b6c4", "#ffffd9"))


lgd_horizontal <- Legend(col_fun = col_fun, title = "Correlation", at = c(0,  1), 
    direction = "horizontal",
    labels = c("Low",  "High"),
    title_gap = unit(5, "mm"),
    title_gp = gpar(fontsize = 20, fontface = "bold"),
    title_position = "topcenter",
    labels_gp =  gpar(fontsize = 15),
    border = "black",
    legend_height = unit(30, "mm"),
    legend_width = unit(150, "mm"))

grid.rect() 
draw(lgd_horizontal)


lgd_vertical <- Legend(col_fun = col_fun, title = "Correlation", at = c(0,  1), 
    direction = "vertical",
    labels = c("Low",  "High"),
    title_gap = unit(5, "mm"),
    title_gp = gpar(fontsize = 20, font.family = "sans"),
    title_position = "leftcenter-rot",
    labels_gp =  gpar(fontsize = 15),
    # gap = unit(20, "mm"),
    border = "black",
    legend_height = unit(150, "mm"),
    legend_width = unit(150, "mm"))

grid.rect() 
draw(lgd_vertical)


```

```{r}

hm_fig <- image_read("Figures/Article/SampleComparison/palette_9.png")
hm_gg <- image_ggplot(hm_fig)

# leg_fig <- image_read("Figures/Article/SampleComparison/legend_vertical.png")
# leg_gg <- image_ggplot(leg_fig)

fig_1 <- tree_vert
fig_2 <- hm_gg
fig_3 <- tree_horo
# fig_4 <- leg_gg

layout_1 <-  c(

  area(t = 1, l = 0, b = 101, r = 50), # Vertical tree (fig 1)
  area(t = 0, l = 10, b = 100, r = 160), # HM  (fig 2)    
  area(t = 100, l = 45, b = 140, r = 125) # Horizontal tree (fig 3)
  # area(t = 0, l = 0, b = 100, r = 9) # legend

)

fig_1 + fig_2 + fig_3 +   plot_layout(design = layout_1) 
# Fig dim: 1700 x 1000


# Density plots

p1+p2+p3
p1/plot_spacer()/p2/plot_spacer()/p3 + plot_layout(heights = c(1, 0.5, 1, 0.5, 1))

p1/plot_spacer()|p2/p3
# p1_ridge/p2_ridge/p3_ridge



```





























